src/fs.rs

```diff
@@ -0,0 +1,419 @@
+        assert!(lock_type & LOCK_UN == 0, "Do not use lock_file to unlock");
+
+        if path_get_data_remoteness() == DirRemoteness::remote {
+            return false;
+        }
+
+        loop {
+            if unsafe { flock(file.as_raw_fd(), lock_type) } != -1 {
+                return true;
+            }
+            if errno::errno().0 != EINTR {
+                return false;
+            }
+        }
+    }
+
+    /// Creates a `LockedFile`.
+    /// Use this for any access to a which requires mutual exclusion, as it ensures correct locking.
+    /// Use `LockingMode::Exclusive` if you want to modify the file in any way.
+    /// Otherwise you should use `LockingMode::Shared`.
+    /// `file_name` should just be a name, not a full path.
+    /// `file_flags` and `file_mode` are used for opening (and potentially creating) the file.
+    /// Note that this tries to create a separate lock file in the same directory as the file.
+    pub fn new(
+        locking_mode: LockingMode,
+        file_path: &wstr,
+        file_flags: OFlag,
+        file_mode: nix::sys::stat::Mode,
+    ) -> std::io::Result<Self> {
+        const LOCK_FILE_SUFFIX: &wstr = L!(".lock");
+        let mut lock_file_path = file_path.to_owned();
+        lock_file_path.push_utfstr(LOCK_FILE_SUFFIX);
+        let lock_file_path = lock_file_path;
+
+        // We start by locking the dedicated lock file.
+        // This is required to avoid racing modifications by other threads/processes.
+        // If the lock is not released explicitly, it is released when the file descriptor of
+        // `lock_file` is closed, which happens in the `Drop` implementation of `File`.
+        let mut lock_file = wopen_cloexec(
+            &lock_file_path,
+            OFlag::O_RDONLY | OFlag::O_CREAT,
+            Mode::S_IRUSR,
+        )?;
+        unsafe {
+            Self::maybe_lock_file(&mut lock_file, locking_mode.to_flock_op());
+        }
+
+        // Open the data file
+        let data_file = wopen_cloexec(file_path, file_flags, file_mode)?;
+
+        Ok(Self {
+            data_file,
+            _lock_file: lock_file,
+        })
+    }
+
+    pub fn get_file(&self) -> &File {
+        &self.data_file
+    }
+
+    pub fn get_file_mut(&mut self) -> &mut File {
+        &mut self.data_file
+    }
+}
+
+/// Runs the `load` function, which should see a consistent state of the file at `path`.
+/// To ensure a consistent state, we use locks to prevent modifications by others.
+/// If locking is unavailable, the `load` function might be executed multiple times,
+/// until it manages a run without the file being modified in the meantime, or until the maximum
+/// number of allowed attempts is reached.
+/// If the file does not exist this function will return an error.
+pub fn lock_and_load<F, U>(
+    path: &wstr,
+    locking_enabled: &RelaxedAtomicBool,
+    load: F,
+) -> std::io::Result<U>
```

Comment:
It would be helpful if the commit message of the last commit was structured by
- first describing the existing state (this is helpful to give the reader some context)
- then describing the intended behavior changes

to make it easy to see quickly what exactly changed
See https://github.com/git/git/blob/8613c2bb6cd16ef530dc5dd74d3b818a1ccbf1c0/Documentation/SubmittingPatches#L300-L305